userspace/libsinsp/async_docker_metadata_source.cpp

```diff
@@ -0,0 +1,349 @@
+/*
+Copyright (C) 2018 Sysdig, Inc.
+
+This file is part of sysdig.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+#include "async_docker_metadata_source.h"
+
+#if defined(CYGWING_AGENT)
+#    include "async_windows_docker_metadata_source.h"
+#else
+#    include "async_linux_docker_metadata_source.h"
+#endif
+
+#include "sinsp_int.h"
+#include "logger.h"
+
+#if defined(LOCAL_DEBUG)
+#       include<cstdio>
+#       define LOG(fmt, ...) fprintf(stderr, "[%s]:%d: " fmt "\n", __FUNCTION__, __LINE__, ##__VA_ARGS__)
+#else
+#       define LOG(fmt, ...) do { } while(false)
+#endif
+
+using namespace sysdig;
+
+async_docker_metadata_source::async_docker_metadata_source(
+		const std::string& api_version,
+		const bool query_image_info):
+	async_metadata_source<std::string, docker_metadata>(NO_LOOKUP_WAIT,
+	                                                    MAX_TTL_MS),
+	m_query_image_info(query_image_info),
+	m_api_version(api_version)
+{ }
+
+const std::string& async_docker_metadata_source::get_api_version() const
+{
+	return m_api_version;
+}
+
+bool async_docker_metadata_source::query_image_info() const
+{
+	return m_query_image_info;
+}
+
+void async_docker_metadata_source::set_query_image_info(const bool query_info)
+{
+	m_query_image_info = query_info;
+}
+
+void async_docker_metadata_source::run_impl()
+{
+	while(queue_size() > 0)
```

Comment:
Yes, there's only one async thread and the async thread is the only thing that can pull things out of the queue -- the API is protected.

I might re-work this so that we don't have to make two API calls.  I think that'll be cleaner and will address your concern here.

Reply:
I've eliminated the `queue_size()` method all together and updated the dequeue method to return true/false if there was an element to dequeue.  This will address any potential future races between the size and the dequeue methods.